src/Evolution/Systems/ScalarWave/Equations.cpp

```diff
@@ -27,6 +30,102 @@ void ComputeDuDt<Dim>::apply(
     dt_phi->get(d) = -d_pi.get(d);
   }
 }
+/// \endcond
+
+template <size_t Dim>
+void ComputeNormalDotFluxes<Dim>::apply(
+    const gsl::not_null<Scalar<DataVector>*> pi_normal_dot_flux,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        phi_normal_dot_flux,
+    const gsl::not_null<Scalar<DataVector>*> psi_normal_dot_flux,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        interface_unit_normal) noexcept {
+  // We assume that all values of psi_normal_dot_flux are the same. The reason
+  // is that std::fill is actually surprisingly/disappointingly slow.
+  if (psi_normal_dot_flux->get()[0] != 0.0) {
+    std::fill(psi_normal_dot_flux->get().begin(),
+              psi_normal_dot_flux->get().end(), 0.0);
+  }
+
+  get(*pi_normal_dot_flux) = get<0>(interface_unit_normal) * get<0>(phi);
+  for (size_t i = 1; i < Dim; ++i) {
+    get(*pi_normal_dot_flux) += interface_unit_normal.get(i) * phi.get(i);
+  }
+
+  for (size_t i = 0; i < Dim; ++i) {
+    phi_normal_dot_flux->get(i) = interface_unit_normal.get(i) * get(pi);
+  }
+}
+
+template <size_t Dim>
+void UpwindFlux<Dim>::package_data(
+    const gsl::not_null<Variables<package_tags>*> packaged_data,
+    const Scalar<DataVector>& normal_dot_flux_pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_dot_flux_phi,
+    const Scalar<DataVector>& /*normal_dot_flux_psi*/,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal)
+    const noexcept {
+  // Computes the contribution to the numerical flux from one side of the
+  // interface.
+  //
+  // The packaged_data stores:
+  // <Pi> = 0.5 * pi
+  // <NormalDotFlux<Pi>> = 0.5 * normal_dot_flux_pi
+  // <NormalTimesFluxPi_i> = 0.5 * normal_dot_flux_pi * n_i
+  // <NormalDotFlux<Phi>_i> = 0.5 * normal_dot_flux_phi_i
```

Comment:
Is doing the 0.5 factors this way actually gaining us anything?  It looks like we now do more additional multiplications in the package_data function than we save in operator() and the code is more confusing.  The only case where this could be more efficient is if the elements are stepping at different rates so we reuse the packaged data for multiple calls in the LTS algorithm, but I expect most boundaries will not be split like that.

I worry that this is a micro-optimization for the uncommon case that both slows down the common case and makes the code harder to understand.

Reply:
Alright, I did a benchmark:

```
1/2 is in package_data
bench_upwind/2_mean           352 ns        352 ns    1994242
bench_upwind/2_median         350 ns        350 ns    1994242
bench_upwind/2_stddev           7 ns          7 ns    1994242
bench_upwind/4_mean           694 ns        694 ns    1005246
bench_upwind/4_median         690 ns        690 ns    1005246
bench_upwind/4_stddev          10 ns         10 ns    1005246
bench_upwind/6_mean          2708 ns       2708 ns     260929
bench_upwind/6_median        2699 ns       2700 ns     260929
bench_upwind/6_stddev          22 ns         22 ns     260929
bench_upwind/8_mean          6496 ns       6496 ns     107215
bench_upwind/8_median        6490 ns       6491 ns     107215
bench_upwind/8_stddev          49 ns         49 ns     107215
bench_upwind/10_mean        16767 ns      16768 ns      41589
bench_upwind/10_median      16753 ns      16753 ns      41589
bench_upwind/10_stddev        145 ns        145 ns      41589

1/2 is in call operator
bench_upwind/2_mean           282 ns        282 ns    2494796
bench_upwind/2_median         282 ns        282 ns    2494796
bench_upwind/2_stddev           4 ns          4 ns    2494796
bench_upwind/4_mean           583 ns        583 ns    1170752
bench_upwind/4_median         581 ns        581 ns    1170752
bench_upwind/4_stddev           6 ns          6 ns    1170752
bench_upwind/6_mean          2420 ns       2420 ns     291388
bench_upwind/6_median        2417 ns       2417 ns     291388
bench_upwind/6_stddev          20 ns         20 ns     291388
bench_upwind/8_mean          6089 ns       6089 ns     113148
bench_upwind/8_median        6081 ns       6082 ns     113148
bench_upwind/8_stddev          37 ns         37 ns     113148
bench_upwind/10_mean        15252 ns      15252 ns      45742
bench_upwind/10_median      15248 ns      15249 ns      45742
bench_upwind/10_stddev        134 ns        134 ns      45742
```

The code is below. I use `ClobberMemory()` to emulate that these things will be generally happening out of order and on different CPUs, so caches aren't helpful. Based on this, it's probably worth changing the 0.5 to be in the call operator. I'll do this and push a fixup.

```cpp
static void bench_upwind(benchmark::State& state) {
  const size_t pts_1d = state.range(0);
  constexpr const size_t Dim = 3;
  const Index<Dim> extents(pts_1d);
  CoordinateMaps::AffineMap map1d(-1.0, 1.0, -1.0, 1.0);
  using Map3d = CoordinateMaps::ProductOf3Maps<CoordinateMaps::AffineMap,
                                               CoordinateMaps::AffineMap,
                                               CoordinateMaps::AffineMap>;
  CoordinateMap<Frame::Logical, Frame::Grid, Map3d> map(
      Map3d{map1d, map1d, map1d});

  Variables<typename ScalarWave::UpwindFlux<Dim>::package_tags> packaged_data_0(
      pow<Dim>(pts_1d), 8.7),
      packaged_data_1(pow<Dim>(pts_1d), 8.7);

  Scalar<DataVector>
      normal_dot_flux_pi(pow<Dim>(pts_1d), 8.7),
      normal_dot_flux_psi(pow<Dim>(pts_1d), -8.7),
      pi(pow<Dim>(pts_1d), 2.0);
  tnsr::i<DataVector, Dim, Frame::Inertial> normal_dot_flux_phi(
      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 8.59)),
      unit_normal(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 1.0));

  Scalar<DataVector> half_normal_dot_flux_pi_interior(pow<Dim>(pts_1d), 3.0),
      half_normal_dot_flux_pi_exterior(pow<Dim>(pts_1d), -3.1),
      half_pi_interior(pow<Dim>(pts_1d), 4.1),
      half_pi_exterior(pow<Dim>(pts_1d), -4.2),
      pi_normal_dot_numerical_flux(pow<Dim>(pts_1d), 0.0),
      psi_normal_dot_numerical_flux(pow<Dim>(pts_1d), 0.0);
  tnsr::i<DataVector, Dim, Frame::Inertial> half_normal_dot_flux_phi_interior(
      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 7.89)),
      half_normal_times_flux_pi_interior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi,
                                                                     -9.27)),
      half_normal_dot_flux_phi_exterior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 3.99)),
      half_normal_times_flux_pi_exterior(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi,
                                                                     -2.59)),
      phi_normal_dot_numerical_flux(
          make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(pi, 0.0));

  ScalarWave::UpwindFlux<Dim> flux_computer{};

  while (state.KeepRunning()) {
    benchmark::ClobberMemory();
    flux_computer.package_data(make_not_null(&packaged_data_0),
                               normal_dot_flux_pi, normal_dot_flux_phi,
                               normal_dot_flux_psi, pi, unit_normal);
    benchmark::ClobberMemory();
    flux_computer.package_data(make_not_null(&packaged_data_1),
                               normal_dot_flux_pi, normal_dot_flux_phi,
                               normal_dot_flux_psi, pi, unit_normal);
    benchmark::ClobberMemory();
    flux_computer(make_not_null(&pi_normal_dot_numerical_flux),
                  make_not_null(&phi_normal_dot_numerical_flux),
                  make_not_null(&psi_normal_dot_numerical_flux),
                  half_normal_dot_flux_pi_interior,
                  half_normal_dot_flux_phi_interior, half_pi_interior,
                  half_normal_times_flux_pi_interior,
                  half_normal_dot_flux_pi_exterior,
                  half_normal_dot_flux_phi_exterior, half_pi_exterior,
                  half_normal_times_flux_pi_exterior);
    benchmark::ClobberMemory();
    flux_computer(make_not_null(&pi_normal_dot_numerical_flux),
                  make_not_null(&phi_normal_dot_numerical_flux),
                  make_not_null(&psi_normal_dot_numerical_flux),
                  half_normal_dot_flux_pi_interior,
                  half_normal_dot_flux_phi_interior, half_pi_interior,
                  half_normal_times_flux_pi_interior,
                  half_normal_dot_flux_pi_exterior,
                  half_normal_dot_flux_phi_exterior, half_pi_exterior,
                  half_normal_times_flux_pi_exterior);
  }
}
BENCHMARK(bench_upwind)->Arg(2)->Arg(4)->Arg(6)->Arg(8)->Arg(10);
```